client/components/hooks/UseSearch.tsx

```diff
@@ -0,0 +1,37 @@
+import { useState, useMemo, Dispatch, SetStateAction } from 'react';
+import Fuse from 'fuse.js';
+
+export type SearchItemData = {
+  name: string,
+  description: string
+}
+
+export type SearchItem = { id: number } & SearchItemData;
+
+type UseSearchReturn = [(Fuse.FuseResult<SearchItem> | SearchItem)[], Dispatch<SetStateAction<string>>];
```

Comment:
```suggestion
type SearchResult = (Fuse.FuseResult<SearchItem> | SearchItem)[];
type UseSearchReturn = [SearchResult, Dispatch<SetStateAction<string>>];
```

also is `(Fuse.FuseResult<SearchItem>[] | SearchItem[])` a more accurate type?

Reply:
[seems like the array brackets should be outside the union type ](https://www.codecademy.com/learn/learn-typescript/modules/learn-typescript-union-types/cheatsheet)

Reply:
The distinction here is that the array is either 100% `SearchItem`s (i.e. `formattedSearchItems`) or 100% `Fuse.FuseResult<SearchItem>`s (i.e. `fuse.search(searchValue)`).

`(Fuse.FuseResult<SearchItem> | SearchItems)[]` is a slightly looser type that lets you have both in the same array

![image](https://user-images.githubusercontent.com/30944418/108608011-ee46c680-7391-11eb-875c-7633369605c6.png)

Reply:
Ohhh I see, that makes sense :)

Reply:
Getting error that the union type is not callable :( https://stackoverflow.com/questions/49365114/typescript-fails-type-resolution-on-union-type. Maybe the way we had it before is okay?

Reply:
😩 TypeScript

Yeah the way it was before is best then